### PR TITLE
Remove the old copy propagation pass

### LINKED ISSUE #36673: MIR CopyPropagation is slow
Hello everyone

I've tried searching for a github issue that reports this, but I couldn't find anything, so I hope this isn't a duplicate. I'm on nightly, since my project uses type macros. Overall, there is a good amount of macro expansion going on, since I build parsers for binary data based on structs.

Compilation time for a 5500 LOC project has been around 7-9 seconds on my Core i7-4770, going a bit up and down with changes in the project, and changes in nightly.

However, today this changed very drastically:

[chrivers@helios]~/git/artemis/oxide #cargo build
   Compiling oxide v0.1.0 (file:///home/chrivers/git/artemis/oxide)
    Finished debug [unoptimized + debuginfo] target(s) in 21.94 secs

This is more than twice as slow as before! Something pretty significant must have happened.

This version is definitely affected:
`rustc 1.13.0-nightly (4f9812a59 2016-09-21)`

Honestly, I'm not quite sure when it worked, but I can say that the version I used yesterday, which was probably no more than 1 week old, worked much better.

### LINKED ISSUE #73717: Compile-time regression with mir-opt-level=2
The following rustc floating point test case used to work just fine with `-Zmir-opt-level=3`, but now even just calling `rustc -Zmir-opt-level=2` on it diverges:
```rust
#![feature(track_caller, stmt_expr_attributes)]

// Poor-man's black-box
#[inline(never)]
fn black_box<T>(x: T) -> T { x }

macro_rules! test {
    ($val:expr, $src_ty:ident -> $dest_ty:ident, $expected:expr) => (
        // black_box disables constant evaluation to test run-time conversions:
        assert_eq!(black_box::<$src_ty>($val) as $dest_ty, $expected,
                    "run-time {} -> {}", stringify!($src_ty), stringify!($dest_ty));

        {
            const X: $src_ty = $val;
            const Y: $dest_ty = X as $dest_ty;
            assert_eq!(Y, $expected,
                        "const eval {} -> {}", stringify!($src_ty), stringify!($dest_ty));
        }
    );

    ($fval:expr, f* -> $ity:ident, $ival:expr) => (
        test!($fval, f32 -> $ity, $ival);
        test!($fval, f64 -> $ity, $ival);
    )
}

macro_rules! common_fptoi_tests {
    ($fty:ident -> $($ity:ident)+) => ({ $(
        test!($fty::NAN, $fty -> $ity, 0);
        test!($fty::INFINITY, $fty -> $ity, $ity::MAX);
        test!($fty::NEG_INFINITY, $fty -> $ity, $ity::MIN);
        // These two tests are not solely float->int tests, in particular the latter relies on
        // `u128::MAX as f32` not being UB. But that's okay, since this file tests int->float
        // as well, the test is just slightly misplaced.
        test!($ity::MIN as $fty, $fty -> $ity, $ity::MIN);
        test!($ity::MAX as $fty, $fty -> $ity, $ity::MAX);
        test!(0., $fty -> $ity, 0);
        test!($fty::MIN_POSITIVE, $fty -> $ity, 0);
        test!(-0.9, $fty -> $ity, 0);
        test!(1., $fty -> $ity, 1);
        test!(42., $fty -> $ity, 42);
    )+ });

    (f* -> $($ity:ident)+) => ({
        common_fptoi_tests!(f32 -> $($ity)+);
        common_fptoi_tests!(f64 -> $($ity)+);
    })
}

macro_rules! fptoui_tests {
    ($fty: ident -> $($ity: ident)+) => ({ $(
        test!(-0., $fty -> $ity, 0);
        test!(-$fty::MIN_POSITIVE, $fty -> $ity, 0);
        test!(-0.99999994, $fty -> $ity, 0);
        test!(-1., $fty -> $ity, 0);
        test!(-100., $fty -> $ity, 0);
        test!(#[allow(overflowing_literals)] -1e50, $fty -> $ity, 0);
        test!(#[allow(overflowing_literals)] -1e130, $fty -> $ity, 0);
    )+ });

    (f* -> $($ity:ident)+) => ({
        fptoui_tests!(f32 -> $($ity)+);
        fptoui_tests!(f64 -> $($ity)+);
    })
}

fn main() {
    common_fptoi_tests!(f* -> i8 i16 i32 i64 u8 u16 u32 u64);
    fptoui_tests!(f* -> u8 u16 u32 u64);
    common_fptoi_tests!(f* -> i128 u128);
    fptoui_tests!(f* -> u128);
}
```

Found through the Miri test suite. Unfortunately we have barely any test coverage on the rustc side for mir-opt-level 2 and higher.

Cc @rust-lang/wg-mir-opt

### LINKED ISSUE #76740: Miscompilation introduced by CopyPropagation
To reproduce the issue:
* Checkout 41dc3942eb33e8e882b6e4782a9bd9d2b8970647 (for reproducibility, but might be unnecessary).
* Enable copy propagation so that it runs on mir-opt-level=1.
* ./x.py test --stage 2 src/test/ui (disable incremental compilation, so that rustc is rebuild completely).
* Observe tests failures: "not all trait items implemented, missing ...".

### PR DESCRIPTION
**Pull request by jonas-schievink**
This pass was added a long time ago, and has not really seen much improvement since (apart from some great work in https://github.com/rust-lang/rust/pull/76569 that unfortunately ran into preexisting soundness issues). It is slow and unsound, and we now have a destination propagation pass that performs a related optimization and could be extended.

Closes https://github.com/rust-lang/rust/issues/36673
Closes https://github.com/rust-lang/rust/issues/73717
Closes https://github.com/rust-lang/rust/issues/76740

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
You'll probably know better than me who a good reviewer here is - I don't have sufficient context I think, so I'll let you reassign.

File: compiler/rustc_mir/src/transform/mod.rs [hunk 1]
```diff
@@ -472,7 +471,6 @@ fn run_optimization_passes<'tcx>(
         &simplify_try::SimplifyArmIdentity,
         &simplify_try::SimplifyBranchSame,
         &dest_prop::DestinationPropagation,
-        &copy_prop::CopyPropagation,
         &simplify_branches::SimplifyBranches::new("after-copy-prop"),
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/mod.rs:474]: this should probably be renamed

File: src/test/mir-opt/copy_propagation_arg.rs [hunk 2]
```diff
@@ -1,39 +0,0 @@
-// Check that CopyPropagation does not propagate an assignment to a function argument
```
Comment (oli-obk) [inline src/test/mir-opt/copy_propagation_arg.rs:1]: Do we have tests for all of these situations for the dest-prop opt?

**Comment by oli-obk** [comment]
lgtm. r=me with nits resolved

**Comment by jonas-schievink** [comment]
(the pass doesn't run by default, so there's no perf impact)

**Comment by camelid** [comment]
@jonas-schievink Could you resolve the merge conflicts?

File: src/test/mir-opt/copy_propagation_arg.rs (hunk 2 above)
Comment (jonas-schievink) [reply src/test/mir-opt/copy_propagation_arg.rs:1]: I ported this test over, the results look correct

File: compiler/rustc_mir/src/transform/mod.rs (hunk 1 above)
Comment (jonas-schievink) [reply compiler/rustc_mir/src/transform/mod.rs]: Renamed to `"final"`

**Pull request merged**